### ISSUE TITLE
disallow ambiguous or-patterns under guards

### ISSUE THREAD

**Issue by backtracking** [ocaml-gospel/gospel#192]
On such a pattern the OCaml compiler emits warning 57:
```
let f = function
  | (x, 0) | (1, x) when x = 0 -> 3
  | _ -> 4
```
that is,
```
Warning 57: Ambiguous or-pattern variables under guard;
variable x may match different arguments. (See manual section 9.5)
```
See https://v2.ocaml.org/releases/4.14/htmlman/patterns.html#sss:pat-or

Suggestion: do not allow such patterns in Gospel.

**Comment by mariojppereira** [comment]
Are you suggesting to completely disallow such patterns, or only if there is a `when` clause ?

What is the current status of `when` inside Gospel patterns ?

**Comment by backtracking** [comment]
Only when there is a `when` clause combined with a or-pattern.

**Comment by backtracking** [comment]
Regarding `when`, @paulpatault is currently adding exhaustiveness checks for patterns in Gospel, and this will include `when` guards as well.

**Comment by mariojppereira** [comment]
Great to hear `when` clauses are coming to Gospel :D

**Comment by mariojppereira** [comment]
> On such a pattern the OCaml compiler emits warning 57:
> 
> ```
> let f = function
>   | (x, 0) | (1, x) when x = 0 -> 3
>   | _ -> 4
> ```
> 
> that is,
> 
> ```
> Warning 57: Ambiguous or-pattern variables under guard;
> variable x may match different arguments. (See manual section 9.5)
> ```
> 
> See https://v2.ocaml.org/releases/4.14/htmlman/patterns.html#sss:pat-or
> 
> Suggestion: do not allow such patterns in Gospel.

**Comment by mariojppereira** [comment]
> Only when there is a `when` clause combined with a or-pattern.

I agree with your initial suggestion to simply disallow such patterns in Gospel.
